app/services/broadcasts/welcome_notification/generator.rb

```diff
@@ -11,17 +11,30 @@ def self.call(*args)
       end
 
       def call
-        # This method should find the user based on the `receiver_id`.
-        # It should then determine the appropriate Broadcast for a user,
-        # based on the `created_at` and the different conditions for sending a notification.
-        # `welcome_broadcast = ...`
+        return if commented_on_welcome_thread? || received_notification?
 
-        # Once it has the appropriate Broadcast to be sent, it should send a notification for it:
-        # `Notification.send_welcome_notification(receiver_id, welcome_broadcast.id)`
+        Notification.send_welcome_notification(receiver_id, welcome_broadcast.id)
+      end
+
+      def received_notification?
+        Notification.exists?(notifiable: welcome_broadcast, user: user)
+      end
+
+      def commented_on_welcome_thread?
+        welcome_threads = Article.where("title LIKE 'Welcome Thread - %'").order(created_at: :desc).first
+        Comment.where(commentable: welcome_threads, user: user).any?
       end
 
       private
 
+      def welcome_broadcast
+        @welcome_broadcast ||= Broadcast.find_by(title: "Welcome Notification: welcome_thread")
+      end
+
+      def user
+        @user ||= User.find_by(id: receiver_id)
+      end
```

Comment:
A possible improvement to get rid of those cached objects

```ruby
      def initialize(receiver_id)
        @user = User.find(receiver_id)
        @welcome_broadcast = Broadcast.find_by!(title: "Welcome Notification: welcome_thread")
      end
```

and then 

```ruby
      private

      attr_reader :user, :welcome_broadcast
```

`receiver_id` wouldn't be needed anymore

ps. I used `find` and `find_by` because I guess we want the `Generator` to exit quickly if the user doesn't exist or if the welcome_broadcast doesn't either. Feel free to keep them as is if I'm mistaken

Reply:
This is a great refactor(😄), although I do agree with @vaidehijoshi in that this might not be the best idea for the `welcome_broadcast` yet. I've initialized `user` locally and will push the change up for it shortly!